src/Microsoft.AspNetCore/CertificateLoader.cs

```diff
@@ -0,0 +1,167 @@
+        }
+
+        private abstract class CertificateSource
+        {
+            public string Source { get; set; }
+
+            public abstract X509Certificate2 Load();
+        }
+
+        private class CertificateFileSource : CertificateSource
+        {
+            private readonly string _password;
+
+            public CertificateFileSource(string password)
+            {
+                _password = password;
+            }
+
+            public string Path { get; set; }
+
+            public override X509Certificate2 Load()
+            {
+                var certificate = TryLoad(X509KeyStorageFlags.DefaultKeySet, out var error)
+                    ?? TryLoad(X509KeyStorageFlags.UserKeySet, out error)
+    #if NETCOREAPP2_0
+                    ?? TryLoad(X509KeyStorageFlags.EphemeralKeySet, out error)
+    #endif
+                    ;
+
+                if (error != null)
+                {
+                    throw error;
+                }
+
+                return certificate;
+            }
+
+            private X509Certificate2 TryLoad(X509KeyStorageFlags flags, out Exception exception)
+            {
+                try
+                {
+                    var loadedCertificate = new X509Certificate2(Path, _password, flags);
+                    exception = null;
+                    return loadedCertificate;
+                }
+                catch (Exception e)
+                {
+                    exception = e;
+                    return null;
+                }
+            }
+        }
+
+        private class CertificateStoreSource : CertificateSource
+        {
+            public string Subject { get; set; }
+            public string StoreName { get; set; }
+            public string StoreLocation { get; set; }
+            public bool AllowInvalid { get; set; }
+
+            public override X509Certificate2 Load()
+            {
+                if (!Enum.TryParse(StoreLocation, ignoreCase: true, result: out StoreLocation storeLocation))
+                {
+                    throw new InvalidOperationException($"Invalid store location: {StoreLocation}");
+                }
+
+                using (var store = new X509Store(StoreName, storeLocation))
+                {
+                    X509Certificate2Collection storeCertificates = null;
+                    X509Certificate2Collection foundCertificates = null;
+
+                    try
+                    {
+                        store.Open(OpenFlags.ReadOnly);
+                        storeCertificates = store.Certificates;
+                        foundCertificates = storeCertificates.Find(X509FindType.FindBySubjectName, Subject, validOnly: !AllowInvalid);
+                        var foundCertificate = foundCertificates
+                            .OfType<X509Certificate2>()
+                            .Where(certificate => certificate.Subject.Equals(Subject, StringComparison.Ordinal))
```

Comment:
Yeah, then he learned he was both right and wrong.  Silly @bartonjs.  You want FindByDistinguishedSubjectName, that leaves all of the RDN keys in place, and matches on String.Equals.  So it does this exact check for you, and you can then drop the Where.

(FindBySubjectName drops all the tags out, for... reasons?, and also doesn't reverse the fields.  So where you want to match "CN=localhost, OU=fun times, C=ZZ" it matches "ZZ, fun times, localhost")

FindByDistinguishedName matches IgnoreCase for some reason, so if you want to be case-respecting you might want to leave this here after all.